scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TFRecordSpec.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2017 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.tensorflow
+
+import com.spotify.scio.ScioContext
+import com.spotify.scio.tensorflow.FeatureKind._
+import com.spotify.scio.values.SCollection
+import io.circe.generic.auto._
+import io.circe.syntax._
+import org.apache.beam.sdk.io.Compression
+
+/** Contains metadata about TFRecords. Useful to read the records later on. */
+trait TFRecordSpec {
+  val LATEST_VERSION: Int = 1
+  def toJson(sc: ScioContext, compression: Compression): SCollection[String]
+}
+
+/** Information necessary to extract a given feature in TF. */
+final case class FeatureInfo(name: String, kind: FeatureKind, tags: Map[String, String])
+
+/** TFRecord metadata that will end up serialized. */
+final case class TFRecordSpecConfig(version: Int,
+                                    features: Seq[FeatureInfo],
+                                    compression: Compression) {
+
+  // Circe struggles with Java Enum, the easiest workaround is to turn all Enums into Strings.
+  case class Str(version: Int,
```

Comment:
Probably better with a custom codec and remove this intermediate type? https://circe.github.io/circe/codec.html

Reply:
That's how I started but turned out that using this case class was less code & more simple. I'll give the custom encoder another try and see how it looks.